metadata/src/main/java/org/apache/kafka/controller/LocalLogManager.java

```diff
@@ -0,0 +1,801 @@
+                return;
+            }
+        }
+    }
+
+    /**
+     * The information that is stored inside the leader file.
+     */
+    public static class LeaderInfo {
+        private final int nodeId;
+        private final long epoch;
+
+        public LeaderInfo(int nodeId, long epoch) {
+            this.nodeId = nodeId;
+            this.epoch = epoch;
+        }
+
+        public int nodeId() {
+            return nodeId;
+        }
+
+        public long epoch() {
+            return epoch;
+        }
+
+        static LeaderInfo read(ByteBuffer buf) {
+            int nodeId = buf.getInt();
+            long epoch = buf.getLong();
+            return new LeaderInfo(nodeId, epoch);
+        }
+
+        int size() {
+            return Integer.BYTES + Long.BYTES;
+        }
+
+        void write(ByteBuffer buf) {
+            buf.putInt(nodeId);
+            buf.putLong(epoch);
+        }
+
+        @Override
+        public String toString() {
+            return "LeaderInfo(nodeId=" + nodeId + ", epoch=" + epoch + ")";
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(nodeId, epoch);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (!(o instanceof LeaderInfo)) return false;
+            LeaderInfo other = (LeaderInfo) o;
+            return other.nodeId == nodeId && other.epoch == epoch;
+        }
+    }
+
+    /**
+     * The LeadershipClaimer continuously tries to claim the leadership of the log.
+     * It always needs to be ready to receive InterruptedExceptions, which somewhat
+     * limits what it can do.  For example, we don't want to write in this thread
+     * since if we get interrupted in the middle, we might have a partial write.
+     */
+    class LeadershipClaimerThread extends KafkaThread {
+        private final String leaderPath;
+        private final FileChannel leaderChannel;
+        private volatile boolean shuttingDown = false;
+        private volatile long claimedEpoch = -1;
+
+        LeadershipClaimerThread(Path leaderPath,
+                                FileChannel leaderChannel,
+                                String threadNamePrefix) {
+            super(threadNamePrefix + "LeadershipClaimerThread", true);
+            this.leaderPath = leaderPath.toString();
+            this.leaderChannel = leaderChannel;
+        }
+
+        @Override
+        public void run() {
+            try {
+                log.debug("starting LeadershipClaimerThread");
+                while (!shuttingDown) {
+                    LOCK_REGISTRY.lock(leaderPath, leaderChannel,
+                        () -> {
+                            try {
+                                claimedEpoch = scribeThread.blockingClaim();
+                                while (!shuttingDown) {
+                                    synchronized (this) {
+                                        this.wait();
```

Comment:
Can you avoid the spotbugs suppression by simply checking `if (!shuttingDown) { // avoid spotbugs warning` inside the `synchronized` block?

Reply:
I see where you're coming from here but I'd rather not do that.  It would create confusion about whether shuttingDown needs to be protected by the lock (it doesn't)